### PR TITLE
Add parameter

### PR DESCRIPTION
**Pull request by sowu880**
Add parameter in configuration file.
{
  "processingErrors": "skip"
}
The value could be skip or raise.

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerConfigurations/ProcessingErrorsOption.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Fhir.Anonymizer.Shared.Core.AnonymizerConfigurations
+{
+    public enum ProcessingErrorsOption
+    {
+        raise,
```
Comment (qiwjin) [inline src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerConfigurations/ProcessingErrorsOption.cs:9]: add inline comments here

File: src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/PartitionedExecution/FhirPartitionedExecutor.cs [hunk 2]
```diff
@@ -113,8 +113,15 @@ private async Task<IEnumerable<TResult>> AnonymizeAsync(List<TSource> batchData,
                     try
                     {
                         TResult anonymizedResult = await AnonymizerFunctionAsync(content);
-                        result.Add(anonymizedResult);
-                        batchAnonymizeProgressDetail.ProcessCompleted++;
+                        if(anonymizedResult == null)
```
Comment (qiwjin) [inline src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/PartitionedExecution/FhirPartitionedExecutor.cs:116]: why just use null string here? how about empty string? or nullORempty?

File: src/Microsoft.Health.Fhir.Anonymizer.Shared.CommandLineTool/FilesAnonymizerForJsonFormatResource.cs [hunk 3]
```diff
@@ -86,7 +88,7 @@ public async Task<string> FileAnonymize(string fileName)
 
             if (_options.SkipExistedFile && File.Exists(resourceOutputFileName))
             {
-                Console.WriteLine($"Skip processing on file {fileName}.");
+                Console.WriteLine($"Skip processing on file {fileName} since it already exist in destination.");
```
Comment (BoyaWu10) [inline src/Microsoft.Health.Fhir.Anonymizer.Shared.CommandLineTool/FilesAnonymizerForJsonFormatResource.cs:91]: Typo: exists

File: src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerConfigurations/ProcessingErrorsOption.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+        skip,
```
Comment (BoyaWu10) [inline src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerConfigurations/ProcessingErrorsOption.cs:10]: Maybe capitalize?

File: src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs [hunk 5]
```diff
@@ -65,17 +68,34 @@ public static AnonymizerEngine CreateWithFileContext(string configFilePath, stri
...
         {
             EnsureArg.IsNotNull(element, nameof(element));
+            try
+            {
+                ElementNode resourceNode = ElementNode.FromElement(element);
+                return resourceNode.Anonymize(_rules, _processors);
+            }
+            catch(Exception ex)
+            {
+                if(!(ex is AnonymizerConfigurationErrorsException) && _configurationManager.Configuration.processingErrors == ProcessingErrorsOption.skip)
```
Comment (BoyaWu10) [inline src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs:78]: We can catch AnonymizerConfigurationErrorsException separately to make it more clear.

Comment (sowu880) [reply src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs]: Fixed it, thanks!

File: src/Microsoft.Health.Fhir.Anonymizer.Shared.FunctionalTests/ResourceTests.cs [hunk 6]
```diff
@@ -88,6 +89,22 @@ public void GivenAPatientResource_WhenAnonymizingWithMultipleSubstituteConfig_Th
...
+        }
+
+        [Fact]
+        public void GivenAPatientResource_WhenAnonymizingWithProcessingError_IfRaise_ExceptionWillBeThrown()
+        {
+            AnonymizerEngine engine = new AnonymizerEngine(Path.Combine("Configurations", "configuration-raise-processing-error.json"));
+            string testContent = File.ReadAllText(ResourceTestsFile("patient-basic.json"));
+            Assert.Throws<AnonymizerProcessFailedException>(() => engine.AnonymizeJson(testContent));
+        }
+
```
Comment (BoyaWu10) [inline src/Microsoft.Health.Fhir.Anonymizer.Shared.FunctionalTests/ResourceTests.cs:107]: Can we add a test when users do not configure "processingErrors"? The default behavior is raise right?

File: src/Microsoft.Health.Fhir.Anonymizer.R4.CommandLineTool/configuration-sample.json [hunk 7]
```diff
@@ -1,5 +1,6 @@
 {
   "fhirVersion": "R4",
+  "processingErrors": "raise",
```
Comment (BoyaWu10) [inline src/Microsoft.Health.Fhir.Anonymizer.R4.CommandLineTool/configuration-sample.json:3]: It seems our config file is case insensitive right?

File: src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs [hunk 8]
```diff
@@ -65,17 +68,38 @@ public static AnonymizerEngine CreateWithFileContext(string configFilePath, stri
...
+            }
+            catch (AnonymizerConfigurationErrorsException)
+            {
+                throw;
+            }
+            catch
+            {
+                if(_configurationManager.Configuration.processingErrors == ProcessingErrorsOption.Skip)
+                {
+                    return null;
```
Comment (tongwu-msft) [inline src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs:84]: is there any way for customer to get the detail error message?

File: src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs [hunk 9]
```diff
@@ -65,17 +68,38 @@ public static AnonymizerEngine CreateWithFileContext(string configFilePath, stri
...
         {
             EnsureArg.IsNotNull(resource, nameof(resource));
 
             ValidateInput(settings, resource);
-            var anonymizedResource = AnonymizeElement(resource.ToTypedElement()).ToPoco<Resource>();
+            var anonymizedResource = AnonymizeElement(resource.ToTypedElement())?.ToPoco<Resource>();
+
+            if(anonymizedResource == null)
+            {
+                return null;
```
Comment (tongwu-msft) [inline src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs:100]: I believe we should let customer aware there's a failure and the error details.

File: src/Microsoft.Health.Fhir.Anonymizer.Shared.CommandLineTool/FilesAnonymizerForJsonFormatResource.cs [hunk 10]
```diff
@@ -62,14 +62,16 @@ public async Task AnonymizeAsync()
             stopWatch.Start();
 
             int completedCount = 0;
+            int skippedCount = 0;
             int failedCount = 0;
             Progress<BatchAnonymizeProgressDetail> progress = new Progress<BatchAnonymizeProgressDetail>();
             progress.ProgressChanged += (obj, args) =>
```
Comment (qiwjin) [inline src/Microsoft.Health.Fhir.Anonymizer.Shared.CommandLineTool/FilesAnonymizerForJsonFormatResource.cs:68]: i feel ProgressChanged is an event should be defined in FhirPartitionedExecutor

File: src/Microsoft.Health.Fhir.Anonymizer.Shared.FunctionalTests/ResourceTests.cs (hunk 6 above)
Comment (sowu880) [reply src/Microsoft.Health.Fhir.Anonymizer.Shared.FunctionalTests/ResourceTests.cs:107]: Added it, thanks.

File: src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs (hunk 9 above)
Comment (sowu880) [reply src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs:100]: After discussion, if customer decide to ignore the exception, the error is no longer exception that we do not need to return error details. And we just return null to indicate skipping it.

File: src/Microsoft.Health.Fhir.Anonymizer.R4.CommandLineTool/configuration-sample.json (hunk 7 above)
Comment (sowu880) [reply src/Microsoft.Health.Fhir.Anonymizer.R4.CommandLineTool/configuration-sample.json:3]: Yes

File: src/Microsoft.Health.Fhir.Anonymizer.Shared.CommandLineTool/FilesAnonymizerForJsonFormatResource.cs (hunk 10 above)
Comment (sowu880) [reply src/Microsoft.Health.Fhir.Anonymizer.Shared.CommandLineTool/FilesAnonymizerForJsonFormatResource.cs:68]: We can refine it in another PR.

File: src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs (hunk 8 above)
Comment (sowu880) [reply src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs:84]: They can  not get details if they skip the exception.

**Review by qiwjin — APPROVED** [review]

**Review by BoyaWu10 — APPROVED** [review]

**Pull request merged**